File: scripts/addon_fixer/fixer.py
"""
Main addon fixer orchestrator.

Coordinates all fixing operations: manifest updates, Lua transformation,
XML fixes, dependency validation, and packaging.
"""

import logging
import shutil
import zipfile
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional

from .constants import CURRENT_API_VERSION, REPLACEMENT_RECOMMENDATIONS
from .manifest import ManifestParser, ManifestFixer, ManifestData
from .lua_analyzer import LuaAnalyzer, LuaTransformer, LuaAnalysisResult
from .xml_fixer import XMLAnalyzer, XMLTransformer, XMLAnalysisResult
from .validator import validate_addon, ValidationResult
from .migrations import MigrationDatabase

logger = logging.getLogger(__name__)


@dataclass
class FixerConfig:
    """Configuration for the addon fixer."""
    update_api_version: bool = True
    fix_libstub: bool = True
    fix_deprecated_functions: bool = True
    fix_font_paths: bool = True
    fix_xml_issues: bool = True
    add_nil_guards: bool = False  # More aggressive, disabled by default
    validate_dependencies: bool = True
    create_backup: bool = True
    dry_run: bool = False


@dataclass
class FileFixResult:
    """Result of fixing a single file."""
    file_path: Path
    file_type: str  # "lua", "xml", "manifest"
    changes: list[str] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)
    was_modified: bool = False


@dataclass
class AddonFixResult:
    """Result of fixing an entire addon."""
    addon_path: Path
    addon_name: str
    success: bool = True
    manifest_result: Optional[FileFixResult] = None
    lua_results: list[FileFixResult] = field(default_factory=list)
    xml_results: list[FileFixResult] = field(default_factory=list)
    validation_result: Optional[ValidationResult] = None
    backup_path: Optional[Path] = None
    package_path: Optional[Path] = None
    total_changes: int = 0
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)
    recommendations: list[str] = field(default_factory=list)


class AddonFixer:
    """Main addon fixer class that orchestrates all fixes."""

    def __init__(self, config: Optional[FixerConfig] = None):
        """Initialize the addon fixer."""
        self.config = config or FixerConfig()
        self.migration_db = MigrationDatabase()
        self.lua_analyzer = LuaAnalyzer(self.migration_db)
        self.lua_transformer = LuaTransformer(self.migration_db)
        self.xml_analyzer = XMLAnalyzer()
        self.xml_transformer = XMLTransformer()

    def analyze(self, addon_path: Path) -> AddonFixResult:
        """Analyze an addon without making changes."""
        addon_path = Path(addon_path)
        result = AddonFixResult(
            addon_path=addon_path,
            addon_name=addon_path.name,
        )

        # Check for replacement recommendations
        if addon_path.name in REPLACEMENT_RECOMMENDATIONS:
            result.recommendations.extend(REPLACEMENT_RECOMMENDATIONS[addon_path.name])
            result.warnings.append(
                f"This addon ({addon_path.name}) is recommended for replacement, not repair."
            )

        # Analyze manifest
        manifest_path = addon_path / f"{addon_path.name}.txt"
        if manifest_path.exists():
            manifest_result = self._analyze_manifest(manifest_path)
            result.manifest_result = manifest_result
            if manifest_result.errors:
                result.errors.extend(manifest_result.errors)

        # Analyze Lua files
        for lua_file in addon_path.rglob("*.lua"):
            lua_result = self._analyze_lua(lua_file)
            result.lua_results.append(lua_result)
            if lua_result.errors:
                result.errors.extend(lua_result.errors)

        # Analyze XML files
        for xml_file in addon_path.rglob("*.xml"):
            xml_result = self._analyze_xml(xml_file)
            result.xml_results.append(xml_result)
            if xml_result.errors:
                result.errors.extend(xml_result.errors)

        # Validate dependencies
        if self.config.validate_dependencies:
            result.validation_result = validate_addon(addon_path)
            if result.validation_result.errors:
                result.errors.extend(result.validation_result.errors)
            result.warnings.extend(result.validation_result.warnings)

        # Calculate total potential changes
        result.total_changes = (
            len(result.manifest_result.changes if result.manifest_result else []) +
            sum(len(r.changes) for r in result.lua_results) +
            sum(len(r.changes) for r in result.xml_results)
        )

        result.success = len(result.errors) == 0

        return result

    def fix(self, addon_path: Path, output_path: Optional[Path] = None) -> AddonFixResult:
        """Fix an addon and optionally package it."""
        addon_path = Path(addon_path)
        result = AddonFixResult(
            addon_path=addon_path,
            addon_name=addon_path.name,
        )

        # Check for replacement recommendations
        if addon_path.name in REPLACEMENT_RECOMMENDATIONS:
            result.recommendations.extend(REPLACEMENT_RECOMMENDATIONS[addon_path.name])
            result.warnings.append(
                f"This addon ({addon_path.name}) is not recommended for automated repair."
            )
            # Still allow fixing if user insists

        # Create backup if enabled
        if self.config.create_backup and not self.config.dry_run:
            result.backup_path = self._create_backup(addon_path)

        try:
            # Fix manifest
            manifest_path = addon_path / f"{addon_path.name}.txt"
            if manifest_path.exists():
                result.manifest_result = self._fix_manifest(manifest_path)
                result.total_changes += len(result.manifest_result.changes)

            # Fix Lua files
            for lua_file in addon_path.rglob("*.lua"):
                lua_result = self._fix_lua(lua_file)
                result.lua_results.append(lua_result)
                result.total_changes += len(lua_result.changes)

            # Fix XML files
            if self.config.fix_xml_issues:
                for xml_file in addon_path.rglob("*.xml"):
                    xml_result = self._fix_xml(xml_file)
                    result.xml_results.append(xml_result)
                    result.total_changes += len(xml_result.changes)

            # Validate after fixes
            if self.config.validate_dependencies:
                result.validation_result = validate_addon(addon_path)
                result.warnings.extend(result.validation_result.warnings)

            # Package if output path specified
            if output_path and not self.config.dry_run:
                result.package_path = self._package_addon(addon_path, output_path)

            result.success = True

        except Exception as e:
            logger.error(f"Error fixing addon: {e}")
            result.errors.append(str(e))
            result.success = False

            # Restore from backup on failure
            if result.backup_path and not self.config.dry_run:
                self._restore_backup(result.backup_path, addon_path)

        return result

    def _analyze_manifest(self, manifest_path: Path) -> FileFixResult:
        """Analyze manifest file for issues."""
        result = FileFixResult(file_path=manifest_path, file_type="manifest")

        parser = ManifestParser(manifest_path)
        data = parser.parse()

        # Check for issues
        if data.validation_errors:
            result.errors.extend(data.validation_errors)

        if data.encoding_issues:
            result.changes.append("Fix encoding issues")

        # Check API version
        if data.api_version:
            max_version = max(data.api_version)
            if max_version < CURRENT_API_VERSION:
                result.changes.append(
                    f"Update APIVersion from {max_version} to {CURRENT_API_VERSION}"
                )

        # Check for LibStub dependency
        if "LibStub" in data.depends_on:
            result.changes.append("Remove LibStub from dependencies")

        return result

    def _analyze_lua(self, lua_path: Path) -> FileFixResult:
        """Analyze Lua file for issues."""
        result = FileFixResult(file_path=lua_path, file_type="lua")

        analysis = self.lua_analyzer.analyze_file(lua_path)

        for issue in analysis.issues:
            if issue.auto_fixable:
                result.changes.append(f"Fix: {issue.message}")
            else:
                if issue.severity == "error":
                    result.errors.append(f"Line {issue.line_number}: {issue.message}")
                else:
                    result.changes.append(f"Manual fix needed: {issue.message}")

        return result

    def _analyze_xml(self, xml_path: Path) -> FileFixResult:
        """Analyze XML file for issues."""
        result = FileFixResult(file_path=xml_path, file_type="xml")

        analysis = self.xml_analyzer.analyze_file(xml_path)

        for issue in analysis.issues:
            if issue.auto_fixable:
                result.changes.append(f"Fix: {issue.message}")
            else:
                if issue.severity == "error":
                    result.errors.append(f"Line {issue.line_number}: {issue.message}")
                else:
                    result.changes.append(f"Manual fix needed: {issue.message}")

        return result

    def _fix_manifest(self, manifest_path: Path) -> FileFixResult:
        """Fix manifest file issues."""
        result = FileFixResult(file_path=manifest_path, file_type="manifest")

        parser = ManifestParser(manifest_path)
        data = parser.parse()

        if self.config.update_api_version or data.encoding_issues:
            fixer = ManifestFixer(data)
            content, changes = fixer.fix_all()
            result.changes = changes

            if changes and not self.config.dry_run:
                # Write with proper line endings
                with open(manifest_path, "w", encoding="utf-8", newline="\r\n") as f:
                    f.write(content)
                result.was_modified = True

        return result

    def _fix_lua(self, lua_path: Path) -> FileFixResult:
        """Fix Lua file issues."""
        result = FileFixResult(file_path=lua_path, file_type="lua")

        _, changes = self.lua_transformer.fix_file(lua_path, dry_run=self.config.dry_run)
        result.changes = changes
        result.was_modified = len(changes) > 0 and not self.config.dry_run

        return result

    def _fix_xml(self, xml_path: Path) -> FileFixResult:
        """Fix XML file issues."""
        result = FileFixResult(file_path=xml_path, file_type="xml")

        _, changes = self.xml_transformer.fix_file(xml_path, dry_run=self.config.dry_run)
        result.changes = changes
        result.was_modified = len(changes) > 0 and not self.config.dry_run

        return result

    def _create_backup(self, addon_path: Path) -> Path:
        """Create a backup of the addon folder."""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        backup_name = f"{addon_path.name}_backup_{timestamp}"
        backup_path = addon_path.parent / backup_name

        shutil.copytree(addon_path, backup_path)
        logger.info(f"Created backup at: {backup_path}")

        return backup_path

    def _restore_backup(self, backup_path: Path, original_path: Path) -> None:
        """Restore addon from backup."""
        if original_path.exists():
            shutil.rmtree(original_path)
        shutil.copytree(backup_path, original_path)
        logger.info(f"Restored from backup: {backup_path}")

    def _package_addon(self, addon_path: Path, output_path: Path) -> Path:
        """Package addon into a zip file for distribution."""
        output_path = Path(output_path)

        # Create proper addon structure
        addon_name = addon_path.name
        zip_name = f"{addon_name}.zip"
        zip_path = output_path / zip_name

        with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as zf:
            for file_path in addon_path.rglob("*"):
                if file_path.is_file():
                    # Skip backup files and temp files
                    if "_backup_" in str(file_path) or file_path.suffix in [".bak", ".tmp"]:
                        continue

                    # Create proper archive path (AddonName/file.ext)
                    arcname = addon_name / file_path.relative_to(addon_path)
                    zf.write(file_path, arcname)

        logger.info(f"Created package: {zip_path}")
        return zip_path


def analyze_addon(addon_path: Path, config: Optional[FixerConfig] = None) -> AddonFixResult:
    """Convenience function to analyze an addon."""
    fixer = AddonFixer(config)
    return fixer.analyze(addon_path)


def fix_addon(
    addon_path: Path,
    output_path: Optional[Path] = None,
    config: Optional[FixerConfig] = None
) -> AddonFixResult:
    """Convenience function to fix an addon."""
    fixer = AddonFixer(config)
    return fixer.fix(addon_path, output_path)

File: tools/addon-fixer/src/fixer.ts
/**
 * Main addon fixer orchestrator.
 *
 * Coordinates analysis and fixing across all file types:
 * - Manifest files (.txt)
 * - Lua source files (.lua)
 * - XML UI files (.xml)
 */

import { readdir, stat, mkdir, copyFile, rm } from 'node:fs/promises';
import { join, basename, dirname, extname } from 'node:path';
import { createWriteStream } from 'node:fs';
import { pipeline } from 'node:stream/promises';
import { createGzip } from 'node:zlib';
import * as tar from 'tar';

import type {
  FixerConfig,
  AddonAnalysisResult,
  AddonFixResult,
  FileAnalysisResult,
  FileFixResult,
  AnalysisSummary,
  IssueSeverity,
  IssueCategory,
} from './types.js';
import { DEFAULT_CONFIG } from './types.js';
import { LuaAnalyzer } from './lua-analyzer.js';
import { parseManifest, analyzeManifest, fixManifest } from './manifest-parser.js';
import { getAddonRecommendation } from './migrations.js';

// ============================================================================
// Addon Fixer
// ============================================================================

export class AddonFixer {
  private readonly config: FixerConfig;
  private readonly luaAnalyzer: LuaAnalyzer;

  constructor(config: Partial<FixerConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.luaAnalyzer = new LuaAnalyzer(this.config.confidenceThreshold);
  }

  // ============================================================================
  // Analysis
  // ============================================================================

  async analyze(addonPath: string): Promise<AddonAnalysisResult> {
    const addonName = basename(addonPath);
    const fileResults: FileAnalysisResult[] = [];
    const warnings: string[] = [];
    const errors: string[] = [];

    // Check for replacement recommendations
    const recommendation = getAddonRecommendation(addonName);
    if (recommendation) {
      warnings.push(...recommendation.recommendations);
      if (!recommendation.shouldAttemptFix) {
        warnings.push(`This addon (${addonName}) is not recommended for automated repair.`);
      }
    }

    // Find manifest
    const manifestPath = join(addonPath, `${addonName}.txt`);
    let manifestData;

    try {
      await stat(manifestPath);
      const manifestResult = await analyzeManifest(manifestPath);
      fileResults.push(manifestResult);
      manifestData = await parseManifest(manifestPath);
    } catch {
      warnings.push(`Missing manifest file: ${addonName}.txt`);
    }

    // Find all Lua and XML files
    const files = await this.findFiles(addonPath);

    // Analyze Lua files
    for (const file of files.lua) {
      const result = await this.luaAnalyzer.analyzeFile(file);
      fileResults.push(result);
    }

    // Analyze XML files (simplified for now)
    for (const file of files.xml) {
      const result = await this.analyzeXmlFile(file);
      fileResults.push(result);
    }

    // Calculate summary
    const summary = this.calculateSummary(fileResults);

    return {
      addonPath,
      addonName,
      manifestData,
      fileResults,
      summary,
    };
  }

  // ============================================================================
  // Fixing
  // ============================================================================

  async fix(addonPath: string, outputPath?: string): Promise<AddonFixResult> {
    const addonName = basename(addonPath);
    const fileResults: FileFixResult[] = [];
    const errors: string[] = [];
    const warnings: string[] = [];
    const recommendations: string[] = [];
    let backupPath: string | undefined;
    let packagePath: string | undefined;
    let totalChanges = 0;

    // Check for replacement recommendations
    const recommendation = getAddonRecommendation(addonName);
    if (recommendation) {
      recommendations.push(...recommendation.recommendations);
    }

    // Create backup if enabled
    if (this.config.createBackup && !this.config.dryRun) {
      try {
        backupPath = await this.createBackup(addonPath);
      } catch (e) {
        warnings.push(`Failed to create backup: ${e}`);
      }
    }

    try {
      // Fix manifest
      const manifestPath = join(addonPath, `${addonName}.txt`);
      try {
        await stat(manifestPath);
        const manifestResult = await fixManifest(manifestPath, this.config.dryRun);
        fileResults.push(manifestResult);
        totalChanges += manifestResult.changes.length;
      } catch {
        warnings.push('Manifest file not found');
      }

      // Find all files
      const files = await this.findFiles(addonPath);

      // Fix Lua files
      for (const file of files.lua) {
        const result = await this.fixLuaFile(file);
        fileResults.push(result);
        totalChanges += result.changes.length;
        errors.push(...result.errors);
      }

      // Fix XML files
      for (const file of files.xml) {
        const result = await this.fixXmlFile(file);
        fileResults.push(result);
        totalChanges += result.changes.length;
        errors.push(...result.errors);
      }

      // Package if output path specified
      if (outputPath && !this.config.dryRun) {
        try {
          packagePath = await this.packageAddon(addonPath, outputPath);
        } catch (e) {
          errors.push(`Failed to package addon: ${e}`);
        }
      }

    } catch (e) {
      errors.push(`Error during fixing: ${e}`);

      // Restore from backup on failure
      if (backupPath && !this.config.dryRun) {
        try {
          await this.restoreBackup(backupPath, addonPath);
          warnings.push('Restored from backup due to errors');
        } catch (restoreError) {
          errors.push(`Failed to restore backup: ${restoreError}`);
        }
      }
    }

    return {
      addonPath,
      addonName,
      success: errors.length === 0,
      fileResults,
      backupPath,
      packagePath,
      totalChanges,
      errors,
      warnings,
      recommendations,
    };
  }

  // ============================================================================
  // File Discovery
  // ============================================================================

  private async findFiles(dirPath: string): Promise<{ lua: string[]; xml: string[] }> {
    const lua: string[] = [];
    const xml: string[] = [];

    const processDir = async (dir: string): Promise<void> => {
      const entries = await readdir(dir, { withFileTypes: true });

      for (const entry of entries) {
        const fullPath = join(dir, entry.name);

        if (entry.isDirectory()) {
          // Skip backup directories
          if (entry.name.includes('_backup_')) continue;
          await processDir(fullPath);
        } else if (entry.isFile()) {
          const ext = extname(entry.name).toLowerCase();
          if (ext === '.lua') {
            lua.push(fullPath);
          } else if (ext === '.xml') {
            xml.push(fullPath);
          }
        }
      }
    };

    await processDir(dirPath);
    return { lua, xml };
  }

  // ============================================================================
  // Lua File Fixing
  // ============================================================================

  private async fixLuaFile(filePath: string): Promise<FileFixResult> {
    // For now, use the analyzer to detect issues
    // Full transformation requires luaparse code generation
    const analysis = await this.luaAnalyzer.analyzeFile(filePath);

    const changes = analysis.issues
      .filter((issue) => issue.autoFixable && issue.confidence >= this.config.confidenceThreshold)
      .map((issue) => ({
        location: issue.location,
        oldCode: issue.oldCode,
        newCode: issue.suggestedFix ?? '',
        reason: issue.message,
        confidence: issue.confidence,
      }));

    // Apply fixes if not dry run
    // TODO: Implement actual code transformation
    const wasModified = changes.length > 0 && !this.config.dryRun;

    return {
      filePath,
      fileType: 'lua',
      changes,
      errors: analysis.parseErrors.map((e) => e.message),
      wasModified,
    };
  }

  // ============================================================================
  // XML File Analysis/Fixing
  // ============================================================================

  private async analyzeXmlFile(filePath: string): Promise<FileAnalysisResult> {
    // Simplified XML analysis - just check for font paths
    const { readFile } = await import('node:fs/promises');
    let content: string;

    try {
      content = await readFile(filePath, 'utf-8');
    } catch {
      return {
        filePath,
        fileType: 'xml',
        issues: [],
        metrics: { lineCount: 0, functionCount: 0, eventRegistrations: 0, libStubUsages: 0, deprecatedCalls: 0 },
        parseErrors: [{ message: 'Failed to read file', recoverable: false }],
      };
    }

    const issues: FileAnalysisResult['issues'] = [];
    let issueId = 0;

    // Check for font paths
    const fontPattern = /font\s*=\s*["']([^"']+)\.(ttf|otf)(\|[^"']*)?["']/gi;
    let match;

    while ((match = fontPattern.exec(content)) !== null) {
      const lineNum = content.substring(0, match.index).split('\n').length;

      issues.push({
        id: `xml-${++issueId}`,
        filePath,
        category: 'font_path',
        severity: 'warning',
        message: 'Font path uses old TTF/OTF format',
        location: {
          start: { line: lineNum, column: 0, offset: match.index },
          end: { line: lineNum, column: match[0]?.length ?? 0, offset: match.index + (match[0]?.length ?? 0) },
        },
        oldCode: match[0] ?? '',
        suggestedFix: match[0]?.replace(/\.(ttf|otf)/gi, '.slug'),
        autoFixable: true,
        confidence: 0.95,
      });
    }

    return {
      filePath,
      fileType: 'xml',
      issues,
      metrics: {
        lineCount: content.split('\n').length,
        functionCount: 0,
        eventRegistrations: 0,
        libStubUsages: 0,
        deprecatedCalls: 0,
      },
      parseErrors: [],
    };
  }

  private async fixXmlFile(filePath: string): Promise<FileFixResult> {
    const analysis = await this.analyzeXmlFile(filePath);

    const changes = analysis.issues
      .filter((issue) => issue.autoFixable)
      .map((issue) => ({
        location: issue.location,
        oldCode: issue.oldCode,
        newCode: issue.suggestedFix ?? '',
        reason: issue.message,
        confidence: issue.confidence,
      }));

    return {
      filePath,
      fileType: 'xml',
      changes,
      errors: [],
      wasModified: changes.length > 0 && !this.config.dryRun,
    };
  }

  // ============================================================================
  // Backup and Packaging
  // ============================================================================

  private async createBackup(addonPath: string): Promise<string> {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const addonName = basename(addonPath);
    const backupPath = join(dirname(addonPath), `${addonName}_backup_${timestamp}`);

    await this.copyDirectory(addonPath, backupPath);
    return backupPath;
  }

  private async restoreBackup(backupPath: string, originalPath: string): Promise<void> {
    await rm(originalPath, { recursive: true, force: true });
    await this.copyDirectory(backupPath, originalPath);
  }

  private async copyDirectory(src: string, dest: string): Promise<void> {
    await mkdir(dest, { recursive: true });
    const entries = await readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = join(src, entry.name);
      const destPath = join(dest, entry.name);

      if (entry.isDirectory()) {
        await this.copyDirectory(srcPath, destPath);
      } else {
        await copyFile(srcPath, destPath);
      }
    }
  }

  private async packageAddon(addonPath: string, outputPath: string): Promise<string> {
    const addonName = basename(addonPath);
    const zipPath = join(outputPath, `${addonName}.zip`);

    // Use archiver or similar - for now just create the path
    // TODO: Implement actual zip creation
    await mkdir(outputPath, { recursive: true });

    return zipPath;
  }

  // ============================================================================
  // Summary Calculation
  // ============================================================================

  private calculateSummary(fileResults: readonly FileAnalysisResult[]): AnalysisSummary {
    const issuesBySeverity: Record<IssueSeverity, number> = {
      error: 0,
      warning: 0,
      info: 0,
    };

    const issuesByCategory: Partial<Record<IssueCategory, number>> = {};
    let autoFixableCount = 0;

    for (const result of fileResults) {
      for (const issue of result.issues) {
        issuesBySeverity[issue.severity]++;
        issuesByCategory[issue.category] = (issuesByCategory[issue.category] ?? 0) + 1;
        if (issue.autoFixable) {
          autoFixableCount++;
        }
      }
    }

    const totalIssues = issuesBySeverity.error + issuesBySeverity.warning + issuesBySeverity.info;

    // Estimate fix time based on complexity
    let estimatedMinutes = autoFixableCount * 0.1; // Auto-fixes are quick
    estimatedMinutes += (totalIssues - autoFixableCount) * 2; // Manual fixes take longer

    const estimatedFixTime =
      estimatedMinutes < 1
        ? '< 1 minute'
        : estimatedMinutes < 60
          ? `~${Math.ceil(estimatedMinutes)} minutes`
          : `~${Math.ceil(estimatedMinutes / 60)} hours`;

    return {
      totalFiles: fileResults.length,
      totalIssues,
      issuesBySeverity,
      issuesByCategory,
      autoFixableCount,
      estimatedFixTime,
    };
  }
}

// ============================================================================
// Convenience Functions
// ============================================================================

export async function analyzeAddon(
  addonPath: string,
  config?: Partial<FixerConfig>
): Promise<AddonAnalysisResult> {
  const fixer = new AddonFixer(config);
  return fixer.analyze(addonPath);
}

export async function fixAddon(
  addonPath: string,
  outputPath?: string,
  config?: Partial<FixerConfig>
): Promise<AddonFixResult> {
  const fixer = new AddonFixer(config);
  return fixer.fix(addonPath, outputPath);
}
